### DISCUSSION TITLE
Suggestions: Material Editor

### DISCUSSION THREAD

**Discussion by arvinmoses** [o3de/o3de#2598] [Ideas]
These are some suggestions for the user interface and overall user experience in Material Editor

The goal of the design is to increase modularity and extensibility of the layer system and also to allow for layers of obfuscation between shader artists, material artists, and finally environment or character artists.

1. Rather than toggling features, allow features to be 'added' like the component system.  This allows material artists to add layers of complexity to the material without being overwhelmed by too many parameters.  Obviously being able to 'add' the features vs having them there by default would be dependent on shader branching.
![image](https://user-images.githubusercontent.com/7208364/127483117-f077eeda-166a-4669-a617-4eac5c34e78f.png)
2.  Allow an option to adjust parameter visibility, and min/max limits for interaction.  This allows material artists to have an extra layer of control to limit final choices from being outside of pbr range.  Visibility options would be set on the parent, and inherited by children, allowing various levels of obfuscation.  A material artist may want the roughness value to be fully exposed, while a sr environment artist may only want the roughness max exposed to mid level or jr env artists.
![image](https://user-images.githubusercontent.com/7208364/127483885-c41788fa-7c52-4674-bd9d-910f90f18589.png)
3.  The last one would be something more on the backend.  Currently it looks like layered materials are prebuilt with a set number of layered materials.  My suggestion is to make the layer system more dynamic.  Layers are just containers that hold a given material.  Therefore any material with this type can be a layer.  Any material with this type can hold any number of layers.  Masking is a feature that can be added to the layers, otherwise, if there is no masking feature added, the layer would simply overwrite the layers underneath.  This allows for a very flexible structure where layers can easily nest - very similar to the prefab/component system that the main engine is using.  I know this last one is a bit more abstract.  Would love to have a conversation to better explain/discuss

**Comment by HogJonny-AMZN** [comment]
Our internal design we have been moving towards is this (which is kind of like sketchfab)
![image-2021-01-28-12-28-56-674](https://user-images.githubusercontent.com/23222931/127518436-9ed5f95f-5649-493d-a8d7-953b81588b92.png)

**Reply by HogJonny-AMZN** [reply]
The entire property group can be toggled on/off, which will remove all of the properties for a better/compact information architecture and presentation.  The other benefit of this is discoverability, as you will see what the available property groups that can be enabled are.  We think this is the best compromise...  that said, we also studied/discussed the idea of treating them like components, so we are aligned on that being a possibility.  But we are already working on the design/improvement I presented above.

**Reply by arvinmoses** [reply]
That totally makes sense.  I think this method works, though I think I still lean a bit more towards the component-ish method to keep parity with the way the scene editor works as well as to limit the data shown.  If a system was using an uber shader - regardless of the ability close out sections, there is still a possibility of a plethora of unused details

**Reply by santorac** [reply]
Generally there are two kinds of property groups: always-on groups and optional groups. The screenshot from @HogJonny-AMZN  might not be the best example, because metalness isn't optional, it's a core part of the lighting model. But there are other groups like Displacement for example where you could disable the feature entirely, and get better performance.

Initially I was thinking the component approach could be viable viable. It would provide the same end result under the hood: an absent component just means the group is disabled. But what do you do with inherited materials? Suppose there is a material A that enables the Displacement group. Then there is some material B that inherits from A, and it disables the Displacement group. Since "disabling" a group is represented as an absent component, how would the UI communicate the fact that material B has overridden the enable state of the Displacement feature? If we use the checkbox approach, then we have some always-present UI element which could have an alternate appearance (highlighted) to indicate an override.

**Reply by arvinmoses** [reply]
RE: optional and always -on that totally makes sense.  Seems like a really good solution

As far as the disabling of a group, I think you could do it to where at that point it grays out rather than fully removes with the toggle.  Almost like decoupling the visibility and the activation of a feature.  IE adding a feature doesn't automatically enable it.

**Reply by santorac** [reply]
Sure, that could work. You'd just wouldn't be allowed to "remove" a component that was added by a parent.

**Reply by santorac** [reply]
By the way, another very simple change that might help is to just show all the groups as collapsed by default.

**Reply by HogJonny-AMZN** [reply]
I just stole the closest UX mock available, which incorrectly used Metalness - sorry.

**Reply by arvinmoses** [reply]
LOL.

Although I think collapsing groups by default is okay, it won't work in systems with an uber shader imo.  It'll still get too crowded.  I'm saying systems with 40 upwards different features.  It keeps the system easy so that all the mats are basically interchangeable, but you can easily swap features.  

If possible and it's not heavy tool work, it'd be nice to have both.  

Is there a primer on tweaking code yet?  I'd like to try messing with it a bit and see how to dive in the system

**Reply by santorac** [reply]
Regarding "all the mats are basically interchangeable": that concept would be quite a bit more involved than what I was saying. It would be relatively easy to make the group visibility mechanism operate in a "component style" but the set of available property groups would be limited based on the Material Type that you selected. Making material types interchangeable where the material type is _based on_ the combination of components chosen by the user would be quite a large effort as it doesn't follow the overall material type system design that is in place right now.

Regarding documentation, there is some basic information about the material system (but sadly I found the links are broken, reported the issue here https://github.com/o3de/o3de/issues/2542), but it needs to be fleshed out more. I have a task in my personal backlog for this.

**Reply by arvinmoses** [reply]
Again I think I'm miscommunicating a bit.  Not really suggesting interchangeable between material types, but that there would be an uber material type that would cover 90% of material needs.  It would just have features that were able to be turned on or off.  But when I've worked with these in the past, features could get very numerous, and being able to add them through a component like system was useful.  Especially having some type of search functionality associated with it.

**Comment by HogJonny-AMZN** [comment]
"Currently it looks like layered materials are prebuilt with a set number of layered materials." Yes, this is partially because Apocalypse needed material blending and this is the first pass we could provide that added the needed features like: vertex colors and texture blend masks, height based blending, etc. and this could be written/authored as standard shader.  That said, I believe we more or less also want exactly what you describe... a Layered material that can reference existing materials but adds the needed mask inputs and controls, etc.

One incremental improvement we have been working towards, is to allow the material inspector panel to be tabbed, so each material layer can be presented in a tab.  So you'd have the primary tab for the layered material controls, a tab for the first material, and tab for the second, etc.

**Reply by arvinmoses** [reply]
That's awesome to hear.  The thing I would avoid is something like Unreal's layered material system.  For them asset type is unique.  You have material layers, materials, and material blends.  This makes it hard to sort through imo.

My current thought for implementation is that standard materials would have a "blend" feature.  The blend feature would have options.  Otherwise you would just make a material, and you could include that material in another as a layer.  You could also nest layered materials.

As far as sorting and viewing - I think a tabbed view is really interesting.  I haven't experienced anything like that yet.  Another thought could be another outline that just shows all the layers.  Clicking on the layer would open it in the inspector panel.  You could select more than one, to open more than one in the panel

**Reply by HogJonny-AMZN** [reply]
Multiple times in the past 'material splatting' has been requested, even for mesh rendering.  This would be another way to blend materials, only the mesh would have multiple draw calls, one for each material splat.  This would require that a individual material have blending/masks/etc like you suggest.  This would probably be some some kind of 'Material Splat Component', so you can assign the multiple materials to splat in a ordered list... things inherently will get complex, if you have a model with 6 different materials, your 'Material Splat Component' now has a lot of information presented.

As far as the material editor, one way to work would be to have 3 materials open in the viewport tabs:
1 - the multi-layer material
2 - the first input material: sand
3 - the second layer: dirt

That would allow you to visually flip between the materials themselves.
When you are viewing the multi-layer material, is when you would then have multiple tabs in the inspector (one for the property settings for each material layer)

I 'think' that when we get to material graphs, piping 3 materials into the input slot of a multi-layer material, probably has a more intuitive presentation of how the data is connected.

**Reply by arvinmoses** [reply]
Sorry - miscommunicated here I think.

I wasn't suggesting multiple mats and draw calls.  Simply a layer system with a blend feature per layer that allowed the material to blend with layers underneath.

I just don't like how Unreal currently does it where material blends are a completely different asset.  I think it over complicates the process overall.  I'd rather just have a feature in the material that could use a mask to blend with layers underneath.

Regarding tabs that totally makes sense.  I haven't tried working like that yet, but it sounds pretty efficient.

**Reply by HogJonny-AMZN** [reply]
I'm just suggesting that splat layer blending is an option that can get to similar results, just a different approach.

Unreal blends being an asset, might be for reasons similar to why we currently have a separate multi-layer material, a single compiled shader/material representation.

The Order 1886 had some interesting material composition:
https://blog.selfshadow.com/publications/s2013-shading-course/rad/s2013_pbs_rad_slides.pdf

But their material layers, still have the 4 layer restriction, using rgb values to mask/blend, so they are collapsing multiple materials down into a single representation.

I guess the way I think of it, is these are tooling abstractions that are an artists convenience.  You want a nicer more intuitive way to work, of course you do.  Tooling is also where a lot of time goes, that doesn't mean we won't do it (as a community), it's just stating the reality... it will take time.  What is currently provided, is what could be provided without the spend on tooling.

**Reply by arvinmoses** [reply]
I think UE overcomplicated their material layer system.  Even now I can do exactly the same thing without the use of material layers at all and simply use blends.  It doesn't lower the instruction count or anything, but it does reduce the complexity of setting up the system for artists, as well as asset management.

I agree that the tooling will take time and again I appreciate the fact that it feels like o3de is approaching this in making the system efficient vs just feature heavy.  Mostly just tossing ideas on how to approach it if/when it comes around.  FWIW I do find that layers will continually grow in popularity and will be the preferred way to work in the near future.

**Reply by santorac** [reply]
I think one of the reasons Unreal probably chose this approach is because blending material layers is quite complex. I definitely experienced this while implementing StandardMultilayerPBR. Blending PBR materials isn't always as simple as applying a mask. Each layer has to be handled differently, with some of them requiring quite different blend operations. (For base color, roughness, etc, it might be a simple mask. For normals you might reorient them instead of lerp them. For clear coat, do you want separate layers blended together or one clear coat over all the other layers? Also, what's the source of the mask: vertex colors, a single multichannel blend mask, separate blend masks for each layer, blend based on the heightmap?)

So for Unreal, having a separate asset to define how to combine material layers makes sense to me, as there are so many use cases and difficult to make a one-size-fits-all solution. (Full disclosure though, I haven't used it myself).

For Atom's StandardMultilayerPBR we tried to focus on supporting a few common use cases, and our expectation is that game teams will need to implement a custom layered material type for other use cases. To that end, I am planning (and have already prototyped) a significant refactor of the material type specification that will allow a lot more flexibility in sharing code between material types. It will still be more manual than Unreal's approach, but should be a significant improvement over what we have now. I plan to write a RFC on GitHub in the next couple weeks to describe that in detail.

**Reply by arvinmoses** [reply]
Awesome.  I'm very much looking to try out the new prototype.  I have some ideas for how I think a layered material system should function.  So far, most of the systems I've built for are pretty rigid, and while usable, I can't get them to really work the way I'm envisioning.  The results are fine, but I feel like they lack some ease of use.  I'm hoping that the more modular and generic approach the o3de team seems to be taking would allow for more of this.

Thank you for updates and hard work.  Will give more feedback when I get the chance.

**Comment by HogJonny-AMZN** [comment]
Additionally, the layered material would somehow reference an existing material ... rather then editing the original material, edits in the layered material itself would probably be treated like a delta/patch/override.  So for instance, adjusting the height bias in the blended material to make it look good, wouldn't inherently change that value on the original material that was referenced (as it was likely already tuned to look good.)

**Reply by arvinmoses** [reply]
Yup this is perfect.  This allows material artists to fully control look dev by only adjusting a few materials.  Proper inheritance would be amazing.  

Also important for the ability of hiding and locking controls.  This way world or asset artists don't ever override things that should not be and thus remove control from lookdev

**Reply by santorac** [reply]
To clarify, we already do have inheritance where one material can inherit another material of the same type. What we don't have yet is inheriting a single layer where a StandardMultilayerPBR material inherits three StandardPBR materials, one for each layer. This will be covered in the RFC that I plan to write in the next couple weeks about material type specification improvements.

**Reply by santorac** [reply]
Regarding the idea of restricting data of child materials, it's interesting and I can see the value in that. At any given material you could mark a property as "final", or replace its min/max range with something more restrictive. Then child materials would not be allowed to operate outside this range. (I think that's what you are saying, right?)

@arvinmoses feel free to write a GitHub issue to track that feature request, and we can consider putting it on our (Amazon's) roadmap. Of course someone else from the community could work on it (this is encouraged!) we should just coordinate to reduce the likelihood of conflicts.

**Reply by arvinmoses** [reply]
Yup referring to this.  My idea for it is that it would basically inherit the limits of the parent.  Parents can adjust, but children have to work within that.  Could be something more like you're saying also.   Not sure how to make a gh issue yet, but will look into it.  tY!

**Reply by santorac** [reply]
There's a "New Issue" button here: https://github.com/o3de/o3de/issues

**Reply by arvinmoses** [reply]
Thank you!  Will add

**Comment by gadams3** [comment]
We have internal support for toggling group visibility. For now, that is primarily used by functors and layers. Leaving the underlying data as is, we could theoretically make it behave more like components on this surface by letting the user toggle the visibility of each group from a drop down. That would not have an impact on the performance or application of that group's properties to the material or shader. While we have discussed it, the material system does not currently support enabling or disabling subsets of a material. There are plans for changing how grouping works to make it more dynamic and support nesting.

**Reply by arvinmoses** [reply]
I'm guessing this is in regard to hiding parameters.  

My assumption is that the controls are decoupled from the actual shader code.  My thought was to simply add another json entry for each property called isVisible, and you could simply set that to true or false.  I may be oversimplifying this however.

**Reply by HogJonny-AMZN** [reply]
Functors are the logical controls between presentation (interface/reflection) and the shader itself, that is the decoupling.  Yes, we would need a way to store the state, that would need to be a part of the material asset (and optimized out of the data for the runtime material.)

**Comment by arvinmoses** [comment]
New Suggestion: Can we view channel buffers in viewports?  Both in material editor and scene editor.  I know we are forward + so we aren't technically buffering these, but would be nice to be able to preview in the viewport.

**Reply by santorac** [reply]
Yes, I have an item in my backlog for this kind of thing.
